src/main.rs

```diff
@@ -42,15 +68,138 @@ async fn main() {
         .value_of("ckb_uri")
         .unwrap_or("http://127.0.0.1:8114")
         .to_owned();
-    if !uri.starts_with("http") {
-        uri = format!("http://{}", uri);
-    }
 
-    let client = http::connect(&uri)
-        .await
-        .expect(&format!("Failed to connect to {:?}", uri));
+    if uri.starts_with("tcp:://") {
+        let uri = uri.split_off(7);
+        let codec = StreamCodec::stream_incoming();
+        let tcp_stream = TcpStream::connect(&uri)
+            .await
+            .expect(&format!("Failed to connect to {:?}", uri));
+        let (sink, stream) = codec.framed(tcp_stream).split();
+        let sink = sink.sink_map_err(|e| error!("tcp sink error: {:?}", e));
+        let stream = stream.map_err(|e| error!("tcp stream error: {:?}", e));
+
+        let (duplex, rpc_channel) = duplex(
+            Box::pin(sink),
+            Box::pin(
+                stream
+                    .take_while(|x| futures::future::ready(x.is_ok()))
+                    .map(|x| x.expect("stream is closed on errors")),
+            ),
+        );
+        tokio::spawn(duplex);
 
-    service.poll(client).await;
+        if index_tx_pool {
+            // subscribe `new_transaction` topic
```

Comment:
It's better to use a single TCP connection which can guarantee the message sequence. Otherwise the client may receive `rejected_transaction` before `new_transaction`, and the transaction will remain in ckb indexer.

Reply:
CKB now persists tx pool during restart. Also ckb-indexer may start when ckb has already been running for a while. Should ckb-indexer also scan the existing transactions as well? This is complex because the client must subscribe to the events and queue the messages while processing the existing transactions.

Reply:
> It's better to use a single TCP connection which can guarantee the message sequence. Otherwise the client may receive `rejected_transaction` before `new_transaction`, and the transaction will remain in ckb indexer.

I can confirm that current implementation of `rpc_channel` is using a single TCP connection.

Reply:
> CKB now persists tx pool during restart. Also ckb-indexer may start when ckb has already been running for a while. Should ckb-indexer also scan the existing transactions as well? This is complex because the client must subscribe to the events and queue the messages while processing the existing transactions.

unlike connecting to ckb http rpc, I cannot find an elegant way to handle tcp rpc re-connection, when ckb node was restarted, ckb-indexer will output the error log like `cannot get block from ckb node, error: send failed because receiver is gone`, user have to restart the ckb-indexer as well.